src/storage/nats_kv.rs

```diff
@@ -0,0 +1,248 @@
+//! Storage engine backed by NATS Kv
+//!
+//! This storage engine enables storing any WADM related information using NATS Kv as a backend
+
+use async_trait::async_trait;
+
+use async_nats::{
+    Client as NatsClient,
+    jetstream::kv::Config as JetstreamKvConfig,
+    jetstream::kv::Store as JetstreamKvStore,
+};
+
+use derive_builder::Builder;
+use semver::{Version, Prerelease, BuildMetadata};
+use serde::{Deserialize, Serialize};
+use serde_json::{
+    from_slice,
+    to_vec,
+    Error as SerdeJsonError,
+};
+use serde::ser::StdError;
+
+use super::{
+    EngineMetadata,
+    LatticeStorage,
+    StorageEngineError,
+    StorageError,
+    StoreOptions,
+};
+
+use super::state::{
+    LatticeState,
+    LatticeStateBuilder,
+    LatticeStateBuilderError,
+};
+
+use crate::storage::{
+    Store,
+    StoreRead,
+    StoreWrite,
+    StoreReadWrite,
+    WithStateMetadata,
+};
+
+const STORAGE_ENGINE_NAME: &str = "nats_kv";
+const STORAGE_ENGINE_VERSION: Version = Version {
+    major: 0,
+    minor: 1,
+    patch: 0,
+    pre: Prerelease::EMPTY,
+    build: BuildMetadata::EMPTY,
+};
+
+////////////
+// Engine //
+////////////
+
+#[derive(Debug)]
+pub struct NatsKvStorageEngine {
+    /// Configuration for the NATS KV storage engine
+    config: NatsKvStorageConfig,
+
+    /// NATS client
+    client: NatsClient,
+}
+
+#[derive(Debug, Builder, Serialize, Deserialize)]
+pub struct NatsKvStorageConfig {
+    /// URL to the nats instance
+    pub nats_url: String,
+
+    /// Prefix to prepend to NATS bucket name
+    pub nats_prefix: Option<String>,
+}
+
+impl NatsKvStorageEngine {
+    #[allow(dead_code)]
+    pub async fn new(config: NatsKvStorageConfig) -> Result<NatsKvStorageEngine, StorageEngineError> {
+        // Connect to NATS
+        let client = async_nats::connect(&config.nats_url).await
+            .map_err(|e| StorageEngineError::Engine(format!("failed to connect: {e}")))?;
+
+        Ok(NatsKvStorageEngine {
+            config,
+            client,
+        })
+    }
+
+    /// Build the top level bucket for the lattice
+    fn build_lattice_bucket_name(&self, lattice_id: String) -> String {
+        let mut bucket = format!("lattice_{lattice_id}");
+        if let Some(prefix) = &self.config.nats_prefix {
+            bucket = format!("{prefix}_{bucket}");
+        }
+        bucket
+    }
+
+    /// Get or create the KV store for NATS jetstream
+    async fn create_kv_store(&self, bucket: String) -> Result<JetstreamKvStore, StorageError> {
+        // Build jetstream client
+        let jetstream = async_nats::jetstream::new(self.client.clone());
+
+        // Create Jetstream KV store
+        let kv = jetstream.create_key_value(JetstreamKvConfig {
```

Comment:
We should upgrade this to check for the existence of the key value bucket. As far as I've been able to test, the only time `create_key_value` passes when an existing bucket is already there is if the configuration is exactly the same.

If we don't do a get first to check for a pre-existing one, then this code will fail if someone has created a key value bucket already with a custom replication and clustering or persistence configuration.

Reply:
From some other code that I've been working on, this attempts to get and, if not found, creates a new one.
```rust

async fn get_or_create_bucket(js: &Context) -> Result<Store> {
    if let Ok(store) = js
        .get_key_value(STATE_BUCKET_NAME)
        .await
        .map_err(|e| RpcError::Nats(e.to_string()))
    {
        Ok(store)
    } else {
        Ok(js
            .create_key_value(KvConfig {
                bucket: STATE_BUCKET_NAME.to_string(),
                description: "my state bucket".to_string(),
                history: 10,
                ..Default::default()
            })
            .await
            .map_err(|e| RpcError::Nats(e.to_string()))?)
    }
}
```

Reply:
I had this earlier on... Can't remember why I removed it. Definitely should be a `get_or_create` rather than a `create`.  Will update this.

- [ ] use get_or_create pattern for bucket